### PR TITLE
perf(oled): replace per-pixel image blit with run-aware HAL primitive

### PR DESCRIPTION
**Pull request by duplico**
## Summary

Part of #261. Replaces the per-pixel image blit in `oled.c` (`gq_draw_image()` / `gq_draw_image_with_mask()`) with a run-aware decode + a new per-platform HAL primitive, `HAL_oled_fill_run(x, y, length, value)`. This is the shared-code half of the fix; the badge implementation is a companion PR against `duplico/qc2024` (`sh1107.c`). Emulator implementation included here (`grlib_gfx_driver.c`).

**Why now:** tonight's first real hardware measurement (`duplico/qc2024#41`, `GQ_PERF_INSTRUMENT` build on an eZ-FET-attached badge) puts a full-screen redraw at **~292 ms**, with `draw_animation_stack()` — RLE decode + the per-pixel blit this PR touches — accounting for **~278 ms (95%)**. That's a ~3.4 FPS ceiling. This is the dominant cost identified in `docs/perf-investigation.md` (suspect #0), now confirmed on real hardware rather than estimated.

## Design

**The old path:** `gq_draw_image()` decoded the RLE stream one pixel at a time (`gq_image_get_pixel()`) and called `Graphics_drawPixelOnDisplay(context->display, x, y, value)` for every pixel — a real function call that itself makes a second, indirect call through `display->callPixelDraw` to the platform's `qc12_oledPixelDraw()`/`gfx_driver_pixelDraw()`. For a full 128x128 background frame that's up to 16,384 two-level call chains, each re-deriving `y/8`/`y%8` bit position math from scratch.

**The new path:**

1. **Run-aware decode** (`oled.c`): `gq_image_get_pixel()` is replaced by a peek/advance pair:
   - `gq_image_peek_run(frame, *out_avail)` — non-mutating; returns the pixel value at the current position and how many consecutive pixels (>= 1) share that value *and* stay within the current image row. A run is never exposed as crossing a row boundary, since a single `HAL_oled_fill_run()` call only ever fills one display row.
   - `gq_image_advance_run(frame, count)` — commits `count` pixels (`1 <= count <= avail` from the immediately preceding peek) and updates decode state (byte/run-boundary prefetch, row wrap) exactly as if the old per-pixel function had been called `count` times in a row. This is what makes the output provably pixel-identical: for any valid `count`, the state transition is the same as running the old loop `count` times, just computed in one step (`pixel_repeat -= count` instead of `count` individual decrements, etc.).

   Peek and advance are separate calls (not one combined "decode next N" call) specifically so `gq_draw_image_with_mask()` can look at *both* the image and mask streams' available run lengths before committing to how far to advance either one.

2. **`HAL_oled_fill_run(x, y, length, value)`** (new HAL primitive, declared in `gamequeer.h`): each RLE run gets exactly one call instead of one call per pixel. Per-platform:
   - **Emulator** (`grlib_gfx_driver.c`): a plain per-column store loop into `uint8_t frame_buffer[x][y]`. Not a perf target — the emulator's pixel write was never the bottleneck (see the "Emulator measurability" note under suspect #0 in `docs/perf-investigation.md`) — this exists for structural correctness and to keep the golden-test suite meaningful.
   - **Badge** (`sh1107.c`, companion PR): precomputes `page = y/8` and `mask = 0x80 >> (y%8)` *once per call* instead of once per pixel, then loops the byte read-modify-write. See that PR for the framebuffer-layout discussion (packed 1bpp, page-major — a horizontal run still needs one byte op per column; the win is eliminating the double-indirect call chain and the redundant per-pixel bit-position math, not fewer byte writes).

3. **Mask/run composition** (`gq_draw_image_with_mask()`): the image and mask are independently RLE-encoded, so their run boundaries don't generally line up even though both share the same width/height (enforced by the caller, `draw_animation_stack()`, via a dimension check before either draw is invoked). Each loop iteration peeks both streams' available run lengths and advances *both* by `min(image_avail, mask_avail)` — the largest batch that's still guaranteed to have one constant image value and one constant mask value throughout. `mask_pixel == 0` (transparent) runs are skipped entirely, same as the old per-pixel path.

4. **Clipping**: the old per-pixel path checked `draw_x >= xMin && draw_x <= xMax && draw_y >= yMin` (note: no `yMax` check — that bound is already enforced structurally by `frame.y_end`, computed in `gq_load_image()`) independently for every pixel. The new path clips once per run by intersecting `[draw_x, draw_x + run_len)` with `[xMin, xMax]` — mathematically equivalent to per-pixel clipping since both the run and the clip bound are contiguous intervals, so this can't produce a different set of drawn pixels. **Scope note, matching the issue's validation-pass correction:** the RLE stream is still decoded in full regardless of clipping — a clipped row/column costs a decode but not a write, exactly as before. Full decode-level row-skipping (skip bytes/runs entirely for off-screen rows without decoding them) isn't feasible without a format change: RLE runs can span multiple rows (a single run's `pixel_repeat` isn't bounded by row width), so there's no way to know how many compressed bytes a given row consumes without walking through it. Out of scope here; noted as a possible follow-up if it's ever worth the format change.

**`Graphics_drawString`/`menu.c`** — explicitly out of scope, per the issue. They share the same per-pixel indirection problem at smaller scale but aren't touched by this PR.

## Operation-count reduction (measured, not estimated)

I instrumented (locally, not committed) a pixel-vs-run counter and ran it against the actual golden fixtures already in this test suite, rather than hand-waving a ratio:

| Fixture | Pass | Pixels | Runs (HAL calls) | Reduction |
|---|---|---|---|---|
| `mask_encoding_a` | unmasked (background) | 1024 (32x32) | 32 | **32x** |
| `mask_encoding_a` | masked (fg+mask) | 1024 | 1024 | 1x (no batching) |
| `mask_encoding_b` | masked (fg+mask), x2 | 1024 each | 1024 each | 1x (no batching) |
| `anim_advance` | unmasked (128x128 bg) | 16,256* | 16,256 | 1x (no batching) |

*16,256 = 127x128, not 128x128 — an existing, unrelated emulator quirk (`OLED_HORIZONTAL_MAX`/`OLED_VERTICAL_MAX` are defined as 127 in `grlib_gfx.h`, one less than `GQ_SCREEN_W`/`GQ_SCREEN_H`), not something this PR introduces or changes.

**Honest takeaway:** the RLE-run-count reduction is highly content-dependent. `mask_encoding_a`'s flat-color background batches 32x; everything else in this small test suite (chosen to exercise masking/animation edge cases, not representative "real" game art) shows *no* batching at all — the dithered/checkerboard-ish content in these fixtures produces run lengths of 1 throughout. The format's structural ceiling is much higher (RLE7 caps a single run at 128 pixels — a whole row), so real flat-color sprite/background art should batch better than these edge-case fixtures do.

Critically, **the win isn't purely the batching.** Even in the 1x (no-batching) case, every call now goes through *one* direct function call (`HAL_oled_fill_run`) instead of *two* (`Graphics_drawPixelOnDisplay` → indirect `display->callPixelDraw` → `qc12_oledPixelDraw`), and the per-pixel `y/8`/`y%8` bit-position math collapses to "once per call" instead of "once per pixel" — both apply unconditionally, regardless of content. The hardware measurement (next section) is what will show the real split between "fewer calls" and "cheaper calls."

## Correctness

All 12 golden/regression `ctest` targets pass **unchanged** (headless build), including `mask_encoding_a`, `mask_encoding_b`, `anim_advance`, and `hello` — no golden PGM was regenerated. Also verified with `GQ_PERF_INSTRUMENT=ON` (13/13, including the new perf/draw-count cross-check) and a plain X11 build (compiles clean, not part of the ctest suite). `clang-format --dry-run --Werror` clean on all changed files.

## What the orchestrator should measure on hardware to prove the win

1. Flash the badge companion PR (`duplico/qc2024`, `sh1107.c`) with `GQ_PERF_INSTRUMENT=1` and re-run the same measurement recipe as `duplico/qc2024#41` (same production self-animating cart, same `perf-harness.sh --run-seconds 0` sampling).
2. Compare `DRAW_ANIMATION_STACK` avg/max against the `~278 ms` baseline. Given the "honest takeaway" above, don't expect a clean 10x from batching alone on all content — the direct-call-vs-indirect-call elimination is the part guaranteed to help regardless of content; actual game art may batch significantly better than this PR's test fixtures did.
3. Recompute the implied FPS ceiling (`1000 / DRAW_OLED_STACK avg`) against the pre-fix ~3.4 FPS baseline.

## Lockstep check

No bytecode/struct/format changes — this is a runtime-internal decode/blit refactor. `gqc` (Python compiler) is unaffected; confirmed no references to `get_pixel`/`HAL_blit`/`fill_run` anywhere under `gqc/src`.

Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>

---

## Update: DRAW_DECODE/DRAW_WRITE perf sections added (head now `9db1cc2`)

Maintainer-requested follow-up on the same branch: added two `GQ_PERF_INSTRUMENT`
sections, `DRAW_DECODE` and `DRAW_WRITE`, wrapping the run-loop's
`gq_image_peek_run()`/`gq_image_advance_run()` pair and the `HAL_oled_fill_run()` call
respectively, in both `gq_draw_image()` and `gq_draw_image_with_mask()`. **Per-run
granularity** (not per-pixel) — the whole point of #261 is that there are few runs, so
timing at run granularity keeps the two extra `HAL_perf_now()` calls per run cheap
relative to the work measured; per-pixel timing would have made the observer effect
comparable to (or larger than) the thing being measured.

**Struct size**: `gq_perf_stats_t` grows from 88 B (5 sections) to **120 B (7
sections)**. Section indices/byte offsets for a raw-memory SBW reader:

| Index | Section | Byte offset (within `gq_perf_stats`) |
|---|---|---|
| 5 | `DRAW_DECODE` | 88 (count), 92 (total_us), 96 (min_us), 100 (max_us) |
| 6 | `DRAW_WRITE` | 104 (count), 108 (total_us), 112 (min_us), 116 (max_us) |

(Indices 0-4 — `DRAW_OLED_STACK`/`DRAW_ANIMATION_STACK`/`HANDLE_EVENTS`/`SYSTEM_TICK`/`OLED_FLUSH` — unchanged from before. Full byte-offset math is spelled out in `gq_perf.h`'s updated struct-layout doc comment.)

**Zero-cost-when-off proof**: sha256 of the built `gamequeer` binary (headless, default `GQ_PERF_INSTRUMENT` off) is **byte-identical** before and after this addition (`495fc97a9320139cf5768eb00810c213924a54b6636caac0fe61464b727ad5fb`), confirming `GQ_PERF_ENTER`/`EXIT`'s `((void)0)` no-op expansion holds at the new call sites too.

**Goldens**: all 12 (13 with `GQ_PERF_INSTRUMENT`) `ctest` targets pass unchanged — the new `GQ_PERF_ENTER`/`EXIT` pairs only wrap existing calls, no pixel-identity logic changed. `clang-format` clean.

**Sanity-checked real output** via `--perf-dump` against `mask_encoding_a`: `draw_decode count=1056`, `draw_write count=528` — matches that fixture's two masked-draw calls' combined run counts (32 + 1024 = 1056 decode calls; roughly half resulted in an actual framebuffer write, the rest were transparent-masked or clipped).

### CODE REVIEW

**Comment by duplico** [comment]
Badge companion PR: duplico/qc2024#43 (implements `HAL_oled_fill_run()` for `sh1107.c`, verified via CLI build with the submodule pinned to this branch — .map size delta and both build flavors reported there).

(AI-generated via Claude Code w/ Claude Fable 5)

File: gamequeer/src/grlib_gfx_driver/grlib_gfx_driver.c [hunk 1]
```diff
@@ -19,6 +20,43 @@ static void gfx_driver_pixelDraw(void *displayData, int16_t x, int16_t y, uint16
...
+    }
+    if (x + length > OLED_HORIZONTAL_MAX) {
+        length = (uint16_t) (OLED_HORIZONTAL_MAX - x);
+    }
+
+    uint8_t pixel = value ? 1 : 0;
+    for (uint16_t i = 0; i < length; i++) {
+        frame_buffer[x + i][y] = pixel;
+    }
+}
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/grlib_gfx_driver/grlib_gfx_driver.c:65]: Fixed in e5ed44d - added an early `return` when `x >= OLED_HORIZONTAL_MAX`, before the clamp arithmetic runs, mirroring the existing `x < 0` guard above it. Verified with an ASan negative-control build: the pre-fix code segfaults on an out-of-bounds write at this exact line when called with x past the right edge (e.g. x=200), while the fixed code returns cleanly. Full ctest suite (both headless and headless+perf-instrument flavors, 12/12 and 13/13) still passes unchanged.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/src/oled.c [hunk 2]
```diff
@@ -70,38 +71,79 @@ void gq_image_load_byte(gq_image_frame_on_screen *frame) {
...
+ * gq_image_peek_run() call on the same frame) and updates decode state
+ * exactly as if gq_image_get_pixel() (the old per-pixel function) had been
+ * called `count` times in a row -- byte/run-boundary prefetch and row-wrap
+ * happen at the same point in the stream either way, so output is
+ * pixel-identical to the old code for any valid `count`.
+ */
+uint8_t gq_image_peek_run(gq_image_frame_on_screen *frame, uint16_t *out_avail) {
+    // NB: gate this function call on gq_image_done() to avoid undefined behavior.
+    // Also, you need to bootstrap it by calling gq_image_load_byte() (done by
+    // gq_load_image()).
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/oled.c]: Good catch — fixed in 0eab174. The comment was ambiguous and could be misread as "call this once gq_image_done() is true." Reworded to state the actual guard explicitly (only call while !gq_image_done(frame)) and pointed at the two real call sites (gq_draw_image()/gq_draw_image_with_mask(), both of which already gate correctly with while (!gq_image_done(...))). Comment-only change — pixel output unaffected; verified with clang-format and ctest 12/12 (headless) + 13/13 (headless+perf-instrument), both green.

(AI-generated via Claude Code w/ Sonnet 5)

File: gamequeer/src/oled.c [hunk 3]
```diff
@@ -70,38 +71,81 @@ void gq_image_load_byte(gq_image_frame_on_screen *frame) {
...
+ *
+ * gq_image_advance_run() commits `count` pixels (1 <= count <=
+ * the avail value returned by the immediately preceding
+ * gq_image_peek_run() call on the same frame) and updates decode state
+ * exactly as if gq_image_get_pixel() (the old per-pixel function) had been
+ * called `count` times in a row -- byte/run-boundary prefetch and row-wrap
+ * happen at the same point in the stream either way, so output is
+ * pixel-identical to the old code for any valid `count`.
+ */
+uint8_t gq_image_peek_run(gq_image_frame_on_screen *frame, uint16_t *out_avail) {
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/oled.c]: Fixed in f43b1d6 - gq_image_peek_run() is only called from within oled.c (no header prototype, no external call sites), so it's now `static`.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/src/oled.c [hunk 4]
```diff
@@ -70,38 +71,81 @@ void gq_image_load_byte(gq_image_frame_on_screen *frame) {
...
+
+    uint16_t avail = frame->pixel_repeat;
+    if (avail > row_remaining) {
+        avail = row_remaining;
+    }
+    *out_avail = avail;
+    return frame->pixel_value;
+}
+
+void gq_image_advance_run(gq_image_frame_on_screen *frame, uint16_t count) {
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/oled.c]: Fixed in f43b1d6 - gq_image_advance_run() has the same profile (no header prototype, only called from oled.c's decode loops right after a paired peek), so it's now `static` too.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/include/gamequeer.h [hunk 5]
```diff
@@ -311,6 +311,39 @@ void led_tick();
...
+ *     write with no further indirection between it and the hardware/emu
+ *     buffer.
+ *   - `length` is always >= 1 when called; a run never crosses a display
+ *     row, so a single call never needs to wrap.
+ *   - The whole run shares one pixel value (0 or 1) by construction --
+ *     see gq_image_peek_run()'s doc comment in oled.c.
+ *
+ * Per-platform implementations: sh1107.c (badge, packed 1bpp page-major
+ * frame_buffer) and grlib_gfx_driver.c (emulator, uint8_t[128][128]
+ * frame_buffer). The badge implementation precomputes the page index and
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/include/gamequeer.h]: Fixed in f43b1d6 - the doc comment now says uint8_t[OLED_HORIZONTAL_MAX][OLED_VERTICAL_MAX] (currently [127][127], exclusive bounds) instead of the stale hardcoded [128][128], matching grlib_gfx_driver.c's actual frame_buffer declaration.

(AI-generated via Claude Code w/ Fable 5)
